Add support for Exasol

kevang opened the pull request:
### CATEGORY

Choose one

- [ ] Bug Fix
- [✓] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Exasol describes itself as a "high-performance, cloud-first analytics database", which could be a good addition to Superset's arsenal of databases. sqlalchemy_exasol (https://github.com/blue-yonder/sqlalchemy_exasol), which uses ODBC, is the sqlalchemy dialect that we used. A change in the SupersetDataFrame class is necessary to get access to data via the SQL Lab.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TEST PLAN
<!--- What steps should be taken to verify the changes -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [✓] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

Commit:
Add support for Exasol

villebro commented on superset/db_engine_specs/exasol.py:
```diff
@@ -0,0 +1,27 @@
+
+    engine = "exa"
+    max_column_name_length = 128
+    odbc_row_results = True
```
When adding new properties, you should add a default value (`False`) to `BaseEngineSpec` in `base.py`.

villebro replied:
At a minimum, all engine specs should define at least a minimum set of time grains. Take a look at e.g. `sqlite.py`.

kevang replied:
Thanks, I'll check and update accordingly.

villebro replied:
Btw, If memory serves me right Exasol is inherently an Oracle dialect, so you might be best off extending the Oracle spec.

kevang replied:
Afaik this is not the case, Exasol borrows heavily from postgres (e.g. `DATE_TRUNC` which I used to implement the time grain functions is the same in both), but is not a clear descendant in the way that for example Redshift is.

villebro commented on superset/dataframe.py:
```diff
@@ -119,6 +119,8 @@ def __init__(self, data, cursor_description, db_engine_spec):
             self.df = pd.DataFrame(data, columns=column_names)
         else:
+            if getattr(db_engine_spec, "odbc_row_results", None):
+                data = [[value for value in row] for row in data]
```
Can you explain why this is needed? Also, if the attribute is added to `BaseEngineSpec`, you should not need to test for it first (see below).

villebro replied:
The problem with fixing this behaviour per engine is that several engines support both odbc and non-odbc drivers (e.g. mssql referenced in the test). Does the default behaviour not work for exa? As pyodbc has been supported for quite some time on other engines, it feels strange that this should need to be handled differently in this case.

kevang replied:
The default behaviour fails during `pd.Dataframe` instantiation with messages similar to `Shape of passed values is (1, 1), indices imply (1, 50)`, or when I `SELECT` only one column (where the two cases converge and the dataframe can be created) it fails at a later stage with `Unserializable [<class 'pyodbc.Row'>]`

While debugging I also tried to leave the data list as is and pass column names like `[["col1"], ["col2"], ["col3"]]` as per the test, but this was not successful either. 

> As pyodbc has been supported for quite some time on other engines, it feels strange that this should need to be handled differently in this case.

🤷‍♂

kevang replied:
Would it be preferable to add a try/expect block instead?

villebro replied:
That still feels hacky, so would prefer not to go down that route. My intuition tells me there might be a bug in the SqlAlchemy connector. Ideally we would like to end up with a solution that works identically for all odbc sources.

mistercrunch replied:
If something special is needed for Exasol/odbc it should be in the engine specific module. One way would be to override `fetch_data` for Exasol.
https://github.com/apache/incubator-superset/blob/master/superset/db_engine_specs/base.py#L225

villebro reviewed (changes requested):
Left some first pass comments.

villebro commented on docs/installation.rst:
```diff
@@ -659,6 +661,25 @@ it in the ``extra`` parameter::
+    EXAHOST = host:8563
+    EXASCHEMA = main
+
+See `SQLAlchemy for Exasol <https://https://github.com/blue-yonder/sqlalchemy_exasol>`_.
```
There is one `https://` too many here

villebro commented:
I tried starting the exasol dev container from here: https://github.com/exasol/docker-db, installed the exasol 6.2.1 driver (MacOS) and the blue-yonder sqlalchemy connector, but was not able to connect Superset to it. Any idea what the `odbc.ini` file should look like on MacOS? Not sure where the `.so` file goes.. Anyway, this is the error I'm getting on Superset: 
![image](https://user-images.githubusercontent.com/33317356/66183668-50f6cf80-e682-11e9-87b2-4e0f56d08d7d.png)

kevang commented:
Many thanks for taking the time to replicate the issue. 

I'm also on Mac and set up superset with http://superset.incubator.apache.org/installation.html#start-with-docker. Inside the container I created `odbc.ini` with the `config_odbc` tool (which, together with the driver can be downloaded from https://docs.exasol.com/connect_exasol/drivers/odbc/odbc_linux.htm). After running the above script, `odbcinst -j` should list `USER DATA SOURCES..: /root/.odbc.ini`. As a final step the contents of `odbc.ini` should be manually changed (after variables are changed accordingly) to

```
root@307bbaccb9d6:/home/superset# cat /root/.odbc.ini
[EXAODBC]
DRIVER = /root/.driver/EXASOL_ODBC-6.1.5/lib/linux/x86_64/libexaodbc-uo2214lv2.so
EXAHOST = host.docker.internal:8563
EXASCHEMA = schema
```

 Then in Python's REPL I can query with:
```
>>> from sqlalchemy import create_engine
>>> e = create_engine("exa+pyodbc://user:password@EXAODBC")
>>> e.connect()
>>> e.execute("select * from table LIMIT 1").fetchall()
```

Commit:
add time grain functions for Exasol

Commit:
remove duplicate of

kevang commented:
I added the `_time_grain_functions` and implemented a try/except solution which at least is driver independent. I understand that this less than ideal, so happy to hear your thoughts.

Commit:
override ExasolEngineSpec's fetch_data method

Commit:
remove duplicate https

villebro commented on superset/db_engine_specs/exasol.py:
```diff
@@ -0,0 +1,48 @@
+
+    @classmethod
+    def fetch_data(cls, cursor, limit: int) -> List[Tuple]:
+        data = super(ExasolEngineSpec, cls).fetch_data(cursor, limit)
```
Nit: this is equivalent to `super().fetch_data(cursor, limit)`.

kevang replied:
Yes ofc, python 2 habits die hard :D Thanks, updated it.

villebro reviewed (changes requested):
This looks like a good solution; other than the nit LGTM!

Commit:
simplify super call

mistercrunch commented:
![yeah](https://user-images.githubusercontent.com/487433/66275910-61ee3f00-e842-11e9-9cfd-a85724578f91.gif)
